File: src/main/java/com/milaboratory/mixcr/cli/postanalysis/CommandPaListMetrics.java
package com.milaboratory.mixcr.cli.postanalysis;

import com.milaboratory.mixcr.basictypes.Clone;
import com.milaboratory.mixcr.cli.ACommandMiXCR;
import com.milaboratory.mixcr.postanalysis.ui.CharacteristicGroup;
import com.milaboratory.util.GlobalObjectMappers;
import picocli.CommandLine;
import picocli.CommandLine.Parameters;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;

@CommandLine.Command(name = "listMetrics",
        sortOptions = false,
        separator = " ",
        description = "List available metrics")
public class CommandPaListMetrics extends ACommandMiXCR {
    @Parameters(description = "Input file with PA results", index = "0")
    public String in;

    @Override
    public void run0() {
        PaResult paResult;
        try {
            paResult = GlobalObjectMappers.PRETTY.readValue(new File(in), PaResult.class);
        } catch (IOException e) {
            throwValidationException("Corrupted PA file.");
            throw new RuntimeException();
        }

        PaResultByGroup result = paResult.results.get(0);
        CharacteristicGroup<Clone, ?>
                biophys = result.schema.getGroup(CommandPaIndividual.Biophysics),
                diversity = result.schema.getGroup(CommandPaIndividual.Diversity);

        for (int i = 0; i < 2; i++) {
            System.out.println();
            CharacteristicGroup<Clone, ?> gr = i == 0 ? biophys : diversity;
            if (i == 0)
                System.out.println("Biophysics metrics:");
            else
                System.out.println("Diversity metrics:");
            result.result.forGroup(gr)
                    .data.values().stream()
                    .flatMap(d -> d.data.values()
                            .stream().flatMap(ma -> Arrays.stream(ma.data)))
                    .map(m -> m.key)
                    .distinct()
                    .forEach(metric -> System.out.println("    " + metric));
        }
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/cli/postanalysis/CommandPaListMetrics.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli.postanalysis

import com.milaboratory.app.InputFileType
import com.milaboratory.app.ValidationException
import com.milaboratory.mixcr.basictypes.Clone
import com.milaboratory.mixcr.cli.MiXCRCommand
import com.milaboratory.mixcr.postanalysis.ui.CharacteristicGroup
import com.milaboratory.mixcr.postanalysis.ui.PostanalysisParametersIndividual.CDR3Metrics
import com.milaboratory.mixcr.postanalysis.ui.PostanalysisParametersIndividual.Diversity
import picocli.CommandLine.Command
import picocli.CommandLine.Parameters
import java.io.IOException
import java.nio.file.Path

@Command(
    description = ["List available metrics"]
)
class CommandPaListMetrics : MiXCRCommand() {
    @Parameters(
        description = ["Input file with PA results"],
        index = "0",
        paramLabel = "input.json[.gz]"
    )
    lateinit var input: Path

    override fun validate() {
        ValidationException.requireFileType(input, InputFileType.JSON, InputFileType.JSON_GZ)
    }

    override fun run0() {
        val paResult: PaResult = try {
            PaResult.readJson(input)
        } catch (e: IOException) {
            throw ValidationException("Corrupted PA file.")
        }
        val result = paResult.results.first()
        println("CDR3 metrics:")
        result.printMetricsForGroup(CDR3Metrics)

        println()

        println("Diversity metrics:")
        result.printMetricsForGroup(Diversity)
    }

    private fun PaResultByGroup.printMetricsForGroup(groupName: String) {
        val group = schema.getGroup<CharacteristicGroup<Clone, *>>(groupName)
        result.forGroup(group).data.values
            .flatMap { chData ->
                chData.data.values.flatMap { metricsArray -> metricsArray.data.toList() }
            }
            .map { metricValue -> metricValue.key }
            .distinct()
            .forEach { metric -> println("    $metric") }
    }
}
